Initial setup (#1)
wswebcreation/wdio-parallel-runner-service: README.md

WebdriverIO Parallel Runner
===========================

> WebdriverIO Parallel Runner Service provides you the option to run each test in a test-file as a single test for optimizing parallelization.
> - This can only be used with Jasmine and Mocha see [TODO](#todo)
> - Needs to follow a certain structure to work, see [What is supported](#what-is-supported)
> - Currently **DOES NOT** support **MULTIPLE** or **NESTED** describes
>
> This module is still in beta and any feedback would be appreciated.

## Why this module
By default WebdriverIO, together with Mocha, Jasmine or Cucumber, can only run tests in parallel based on a test-file,
not based on a single test within a test-file. When you are running your tests in for example a cloud service like Sauce Labs,

> you can get a lot of webdriver commands making it harder to debug. In the below image you see 63 webdriver commands

Reviewer: ```suggestion
you can get a lot of WebDriver commands making it harder to debug. In the below image you see 63 WebDriver commands
```

Author: Tnx

(which is not a lot, but we've seen much more). When the test fails it is hard to determine which command or which test made the
build fail.

> ![More webdirver commands](./docs/assets/all-tests.png)

Reviewer: ```suggestion
![More WebDriver commands](./docs/assets/all-tests.png)
```

Author: Tnx

Secondly you are not optimally using your concurrency, especially if you have for example 100 concurrent sessions which you can use.
In this case you are only using a max of 56.

![Concurrency](./docs/assets/default-tests.png)

When you use this module you can easily debug your tests, take for example the output of the spliced testcase from above.

> ![Less webdirver commands](./docs/assets/single-it.png)

Reviewer: ```suggestion
![Less WebDriver commands](./docs/assets/single-it.png)
```

Author: tnx

And you are using your concurrency to the max

![Optimal concurrency](./docs/assets/single-its.png)

## Installation
The easiest way is to keep `wdio-parallel-runner-service` as a devDependency in your `package.json`.

```json
{
    "devDependencies": {
        "wdio-parallel-runner-service": "^0.1.0"
    }
}
```

You can simple do it by:

```bash
npm install wdio-parallel-runner-service --save-dev
```

Instructions on how to install `WebdriverIO` can be found [here.](https://webdriver.io/docs/gettingstarted.html)

## Configuration

The following code shows the default wdio test runner configuration. Just add `'parallel-runner'` as a service
to the array.

```js
// wdio.conf.js
module.exports = {
  // ...
  services: ['parallel-runner'],
  // ...
};
```

## What is supported
Tests need to have a structure like this.

```js

> import Foo from './foo'; // const Foo = require('foo');

Reviewer: Let's keep it simple and avoid the chance of confusion

```suggestion
const Foo = require('foo');
```

Author: 👍

describe('Foo', () => { // `fdescribe` or `xdecribe` are also supported
    before(() => { // optional
        // Do something
    });

    after(() => { // optional
        // Do something
    });

    beforeEach(() => { // optional
        // Do something
    });

    afterEach(() => { // optional
        // Do something
    });

    beforeAll(() => { // optional
        // Do something
    });

    afterAll(() => { // optional
        // Do something
    });

    it('should be able foo the bar', () => {
        // Do something
    });

    fit('should be able to bar', () => {
        // Do something
    });

    xit('should be able to bar', () => {
        // Do something
    });
});
```

Meaning **ONE** `describe|fdescribe|xdescribe`. If multiple describes are detected the file will be not be split and run in its original state.

The above code will be split into **THREE** files, independent of `it|fit|xit` and will look like this:

```js
// File one

> import Foo from './foo'; // const Foo = require('foo');

Reviewer: ```suggestion
const Foo = require('foo');
```

Author: 👍

describe('Foo (1-3)', () => {
    // all hooks will be copied
    before(() => {});
    after(() => {});
    beforeEach(() => {});
    afterEach(() => {});
    beforeAll(() => {});
    afterAll(() => {});

    it('should be able foo the bar', () => {
        // Do something
    });
});

// File two

> import Foo from './foo'; // const Foo = require('foo');

Reviewer: ```suggestion
const Foo = require('foo');
```

Author: 👍

describe('Foo (2-3)', () => {
    // all hooks will be copied
    before(() => {});
    after(() => {});
    beforeEach(() => {});
    afterEach(() => {});
    beforeAll(() => {});
    afterAll(() => {});

    fit('should be able to bar', () => {
        // Do something
    });
});

// File three:

> import Foo from './foo'; // const Foo = require('foo');

Reviewer: ```suggestion
const Foo = require('foo');
```

Author: 👍

    after(() => {});
    beforeEach(() => {});
    afterEach(() => {});
    beforeAll(() => {});
    afterAll(() => {});

    xit('should be able to bar', () => {
        // Do something
    });
});
```

## Temporary files
As you might have read in [What is supported](#what-is-supported) the test-file is being sliced into multiple files with single tests in them.
To make this work this module will:

- create temporary copies of the original file next to the original file, but with post-fix `{current-of}.js` which will result in:
    `original.spec.js.1-2.js`\
    `original.spec.js.2-2.js`\
    ![Temporary files](./docs/assets/temp-files.png)
- Alter the initial describe name with a post-fix ` (current-of)` resulting in
    `describe('foo (1-2)', ()=>{});`\
    `describe('foo (2-2)', ()=>{});`

The temporary files will be removed when the tests have been executed.

> NOTE: If WebdriverIO breaks somewhere before the test have been finished you might have a bunch of extra files in your folder
> next to your original folder. There is currently no solution for this.

## What is not supported (yet)!
## TypeScript
This module currently doesn't support TypeScript

### Conditional its

> There are situations where you would can a tests like this.

Reviewer: ```suggestion
There are situations where you can have a test like this.
```
